Use cross compile in dockerfile, fix arm32 support

NicolasDorier opened the pull request:
Supersede https://github.com/ElementsProject/lightning/pull/7208

This PR is cross compiling dockerfiles, except for the python dependencies required by cln-rest.

You will find here the difference build stages.
* The blue stages only run on amd64.
* The pink stages run on the target architecture
* The white stages are running on the architecture of the building machine

![docker](https://github.com/ElementsProject/lightning/assets/3020646/bc63d65d-bfbb-41e9-b10b-f6b2676d3cde)

On my computer it takes around 30-40min to build. On github action, it takes 3 hours.

I also simplified `tools/build-release.sh` as to just use buildx the way it is intended to be used (no manual creation of manifest)
Note that it assumes that buildx has been configured before running the script (via `docker buildx create --use`)

NicolasDorier commented:
Tested on amd64 and arm64.

NicolasDorier commented:
Out of the 3H of build, more than 2H is spent building python requirements for cln-rest. Specifically:
* `cryptography` and `gevent` for arm32
* `gevent` for arm64

The build can be a bit shorter if we bump to python at least 3.10 because it is pre-built on pip for `arm64`.

NicolasDorier commented:
Made a small change here https://github.com/ElementsProject/lightning/pull/7214/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R225 so only changes to the cln-rest requirements.txt would invalidate the cached `python-builder`. If you build locally, it means it will rarely have to rebuild `python-builder`, the most time consuming build stage.

rustyrussell reviewed (approved):
This looks like a great cleanup!

Weird that we install tclsh, but that's not new: we always did.

I'm going to get @ShahanaFarooqui to review, since I don't speak Docker...

NicolasDorier commented:
@ShahanaFarooqui I rebased on master and added https://github.com/ElementsProject/lightning/pull/7214/commits/65c3415fe7582988a58726af9e25ddd83f7f27f0

ShahanaFarooqui commented:
ACK 65c3415fe7582988a58726af9e25ddd83f7f27f0

NicolasDorier commented:
@ShahanaFarooqui so I tried to replicate your issue about the need of root, and couldn't do it.

I created a new VM, on debian, removed the `sudo`, and used a non root user.

I had to add this user to the docker group, and tried to build and didn't got the issue.

```
Client: Docker Engine - Community
 Version:           26.1.3
 API version:       1.45
 Go version:        go1.21.10
 Git commit:        b72abbb
 Built:             Thu May 16 08:33:42 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.1.3
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.10
  Git commit:       8e96db1
  Built:            Thu May 16 08:33:42 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.31
  GitCommit:        e377cd56a71523140ca6ae87e30244719194a521
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

ShahanaFarooqui commented:
@NicolasDorier It is strange. My docker version is same as yours and I am in the docker user group as well but it somehow fails with non-root user for me. Just leaving this log below for future debugging purposes. At least it proves that the issue is somewhere related to my user permissions. I will try to debug it one day :D.

Thanks a lot for this pretty neat PR though.

```
ERROR: failed to solve: error from sender: open focal/etc/ssl/private: permission denied

$ groups
shahana adm cdrom sudo dip plugdev lpadmin lxd sambashare docker

$ docker version
Client: Docker Engine - Community
 Version:           26.1.3
 API version:       1.45
 Go version:        go1.21.10
 Git commit:        b72abbb
 Built:             Thu May 16 08:33:29 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.1.3
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.10
  Git commit:       8e96db1
  Built:            Thu May 16 08:33:29 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.31
  GitCommit:        e377cd56a71523140ca6ae87e30244719194a521
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```
